lib/Sema/CSSimplify.cpp

```diff
@@ -1091,14 +1091,16 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
                                      ConstraintLocatorBuilder locator) {
   // An @autoclosure function type can be a subtype of a
   // non-@autoclosure function type.
-  if (func1->isAutoClosure() != func2->isAutoClosure() &&
-      kind < ConstraintKind::Subtype)
-    return SolutionKind::Error;
+  if (func1->isAutoClosure() != func2->isAutoClosure()) {
```

Comment:
Not your bug, but this is wrong in general: This allows passing non-autoclosure closures to autoclosure-typed parameters.

Reply:
Can you please clarify? Please note that:
1) the sole caller of matchFunctionTypes() seems to ensure that non-autoclosure functions are wrapped in autoclosure functions.
2) The assert below should verify that the caller ensured the auto-closure wrapper, right?

Reply:
> the sole caller of matchFunctionTypes() seems to ensure that non-autoclosure functions are wrapped in autoclosure functions.

That's not the way this rule should work.  When matchTypes sees `@autoclosure (T) -> U` and it's trying to perform subtype matching, it should error when it sees another `@autoclosure` parameter and try to match `U` with whatever the second type was covariantly.  Instead the logic for `@autoclosure` functions doing equality matching was merged with subtype matching when trying to fix an older SR.

Reply:
Thanks. I'm new to the code and I don't fully follow, but I think you've given me enough information that I can bootstrap myself.

In any case, is your concern strong enough to not merge this? Should I create a followup bug?

For whatever it may be worth, I'm new to the code, and the different ways of doing subtype checking in matchFunctionTypes() confused me, so I tried to clean it up. Adding the assert makes the auto-closure logic look/smell like the "throws" and "noescape" logic. Especially after the simplification we discussed earlier.

Reply:
I think I understand now. I forced the caller to return an error as you said, and unsurprisingly, the following test failed:

```
test/Constraints/closures.swift:264:33: error: unexpected error produced: cannot convert value of type '((String) -> String)?' to expected argument type '_?'
    let _: (String) -> String = foo ?? bar
                                ^
```

Reply:
Yeah.  I believe it was @gregomni that took that SR.  It needs a more holistic fix.

Reply:
Hi @CodaFi – I've closed this and created a new PR. If you could review #12042, I'd appreciate it. Thanks!